### ISSUE TITLE
Tx Anti-Aggregation for regular (non-Dandelion txs)

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#799]
tldr - 
* Dandelion provides a mechanism for tx aggregation and cut-through.
* We need to support multi-kernel txs for this.
* We need an anti-aggregation mechanism is place for regular non-Dandelion txs (to minimize aggregation abuse on the network).

------

When we support Dandelion we will have the ability for both tx aggregation and tx cut-through at the "stem" phase of Dandelion tx propagation.

Once txs enter the "fluff" phase (or skip Dandelion stem phase entirely for speed) - we want to prevent any further tx aggregation.
If txs are aggregated during normal tx propagation we run a significant risk of conflicts and open the network up to a variety of denial-of-service concerns.

For example - nefarious node starts aggregating random txs in different combinations, preventing other nodes from seeing a consistent view of the current tx pool.
Sends the aggregated tx `(tx1, tx2)` to some peers and a conflicting `(tx1, tx3)` to other peers etc.

The problem - 
If we support Dandelion then we must by definition support txs with multiple kernels.
If we support multi-kernel txs then we have no way of differentiating between the following - 
* cut-through txs
* legitimately aggregated txs (via Dandelion)
* any set of txs aggregated by any node at any point in time, anywhere on the network

This is _good_ for privacy, but _bad_ for the reasons given above.

Proposed solution - 
(This has been discussed a few times previously but putting it down in a specific issue here for reference).
1) allow nodes to aggregate txs (we cannot prevent this)
2) but - if we see evidence of an aggregated tx - 
  1) attempt to de-aggregate it if we have also seen any of the constituent txs
  1) store them locally in our pool as de-aggregated as much as possible
  1) propagate them in this de-aggregated state  as much as possible

Basically if a node sees the following (based on tx kernel sets) -
* `(tx1, tx2)`
* `tx1` (already in pool, accepted prior to aggregated tx)

Then - we can recreate `tx2` by "subtracting" `tx1` from `(tx1, tx2)` (inputs, outputs _and_ kernels).
We then act like we never saw `(tx1, tx2)` but we add `tx2` to our pool and propagate `tx2` out to our peers.

I think this will mean well behaved nodes on the network will actively work to un-aggregate, thus reducing or even eliminating any effect any badly behaved node would have on the network.

This would eliminate the misunderstanding that aggregating txs increases privacy. Arbitrarily aggregating txs adds zero privacy if other nodes have already seen any of the original constituent txs because the unseen "other half" of the tx can be trivially recreated.

Note: This would not prevent a node that originates all individual txs `tx1, tx2, tx3` from combining them in conflicting ways before any other node has seen them. But this problem exists with or without Dandelion. There is nothing preventing a node broadcasting conflicting txs to the network today.

**Comment by quentinlesceller** [comment]
Hey @antiochp, I have a few questions regarding this anti aggregation mechanism:

1) It is only possible to deaggregate a n-kernel transaction if we find k>=n-1 kernels in the transaction pool ?

2) If ```tx1 = A -> B``` and ```tx2 = B -> C```. Then ```(tx1, tx2) = (A-> C)```. If I have ```tx2``` in my mempool. How can I retrieve ```tx1``` in term of rangeproof ?

3) Should we also work on a way to deaggregate ```(tx1, tx2)```, when this multi-kernel transaction is just a "dumb" one (meaning it is basically aggregated without any cut-through).

Let me know if that makes sense.

**Comment by antiochp** [comment]
My understanding of this is we can _only_ de-aggregate "dumb" aggregated txs.
We cannot de-aggregate a cut-through tx as we cannot recreate the cut-through parts of it.
In your cut-through example (2) there is no way to recreate `tx1` based on `(tx1, tx2)` and `tx2` as we do not have `B` anymore. 

The intent here is to minimize (or de-incentivize) the noise created by a badly behaving node aggregating random txs together, by making it trivial for other nodes to "undo" them.

**Comment by quentinlesceller** [comment]
Okay that makes sense. Thanks.

**Comment by antiochp** [comment]
A "dumb" aggregated example - 

```
tx1: A -> B, kernels: K1
tx2: C -> D, kernels: K2
tx1,2: (A, C) -> (B, D), kernels: (K1, K2)
```
Given `tx1,2` and `tx2` we can recreate `tx1`.

If a node can trivially recreate `tx1` then it is not worth a node creating `tx1,2` in the first place.

**Comment by quentinlesceller** [comment]
Curious also, do we have to try all the permutations in order to identify the correct ```inputs/outputs/kernels``` set ?
e.g.:
```
tx1: A -> F, kernels: K1
tx2: C -> D, kernels: K2
tx3: B -> E, kernels: K3

tx1,2,3: (A, B, C) -> (D, E, F), kernels: (K1, K2, K3)

**Comment by antiochp** [comment]
Hmm need to think about this a bit.

If a node has seen `tx1` then it can recreate `tx2,3` given `tx1,2,3`.
If it has seen both `tx1` and `tx2` then it can recreate `tx3`.

I think at each stage it simply attempts to de-aggregate whatever it can.
And maybe the important point is - "it forgets about the aggregated tx and replaces it with the more granular txs".
